spec/support/finders/test_selector.rb

```diff
@@ -0,0 +1,46 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+module TestSelectorFinders
+  def test_selector(value)
+    "[data-test-selector=\"#{value}\"]"
+  end
+
+  def find_test_selector(value, **)
+    find(test_selector(value), **)
+  end
+
+  def within_test_selector(value, **, &block)
+    within(test_selector(value), **, &block)
+  end
+end
+
+RSpec.configure do |config|
+  Capybara::Session.include(TestSelectorFinders)
+  Capybara::DSL.extend(TestSelectorFinders)
+  config.include TestSelectorFinders, type: :feature
+end
```

Comment:
Another way to do it which I read about in this [issue comment](https://github.com/teamcapybara/capybara/issues/2694#issuecomment-1668146851):

```ruby
Capybara.test_id = 'data-test-selector'
Capybara.add_selector(:test_id) do
  xpath do |locator|  
    XPath.descendant[XPath.attr(Capybara.test_id) == locator]    
  end    
end  
```

And then it can be used this way:

```
page.find(:test_id, 'op-timer-account-menu-stop')
page.within(:test_id, 'op-timer-account-menu-stop') { ... }
expect(page).to have_selector(:test_id, 'op-timer-account-menu-stop')
```

It is then possible to define new matchers using this selector:

```ruby
def find_test_selector(...)
  find(:test_id, ...)
end

def within_test_selector(...)
  within(:test_id, ...)
end

def have_test_selector(...)
  Capybara::RSpecMatchers::Matchers::HaveSelector.new(:test_id, ...)
end

page.find_test_selector('op-timer-account-menu-stop')
page.within_test_selector('op-timer-account-menu-stop') { ... }
expect(page).to have_test_selector('op-timer-account-menu-stop')
```

----

Another approach seen in https://github.com/ElMassimo/capybara-compose/blob/main/spec/support/global_filters.rb is to do this:

```ruby
Capybara::Selector.all.each_value do |selector|
  selector.instance_eval do
    expression_filter(:test_selector) do |expression, value|
      builder(expression).add_attribute_conditions(Capybara.test_id => value)
    end
    describe_expression_filters do |**options|
      if (value = options[:test_selector])
        " with data-test-selector #{ value.inspect }"
      end
    end
  end
end
```

It then allows to write this:

```ruby
page.find(test_selector: 'op-timer-account-menu-stop')
page.within(test_selector: 'op-timer-account-menu-stop') { ... }
expect(page).to have_selector(test_selector: 'op-timer-account-menu-stop')
```

I have not tested it thoroughly but I think this can achieve the same effect:

```ruby
  Capybara.modify_selector(:css) do
    filter(:test_selector) { |node, v| node[Capybara.test_id] == v }
  end
```